### PR TITLE
refactor(app, components): add 2d module visualization to def and components, add to protocol setup

### LINKED ISSUE #8175: Accordion step 3: Labware setup follow up
# Overview

Follow up to #7665, wire up call to action button and provide actual tooltip when the logic done.

### LINKED ISSUE #8249: Protocol Upload Revamp: Updated Module visualization component
Updated Module visualization component - SVG

1. incorporate high fidelity svgs into shared react components 
2. organize and deprecate old components

### PR DESCRIPTION
**Pull request by b-cooper**
# Overview

The size of this PR is mostly due to a file organization change in the components library that effects import paths across our JS projects. All hardware simulation related components have been moved from the confusingly named catch-all directory of `components/src/deck` to `components/src/hardware-sim` and `components/src/legacy-hardware-sim`. The only import paths this should effect are one's that dig in from the top level index exported from the components library (e.g. `@opentrons/components/deck/getDeckDefinition.ts`). Apologies for the clutter in the diff, but they should be pretty easy to tell apart from the real code changes.

Closes #8175, Closes #8249

# Changelog

- Incorporate top down 2-dimension SVG renderings of the hardware modules into the module definition JSON schema. 
- Standardize hardware module types in shared-data and create module JSON schema v3 (adds fields necessary for rendering)
- Create new `Module` component that takes a module definition, and positioning information, and renders an SVG based on available data.
- Organize hardware simulation components in library to deprecate unused components and
mark legacy components that are to be phased out. 
- Incorporate new `Module` component into the new protocol setup flow.
- Disable the proceed to run CTA if there are missing modules or unfinished calibration

# Review requests

- [ ] take a look through the code changes and call out anything suspicious
- [ ] inspect the new protocol upload flow.  If the protocol loaded requires modules that are not attached to the current robot, the main "Proceed to Run" CTA should be disabled with an explanatory tooltip
- [ ] inspect the new Module and Labware Setup steps, hifi Module svgs should render in both, though nested labware should only appear in the latter  
- [ ] smoke test PD to ensure that it renders the deck with modules and labware just as it did before
- [ ] smoke test protocol-library kludge to ensure that it's usage of legacy hardware simulation components is uneffected
- [ ] take a look at the new stories associated with the `Module` component in the components library

# Risk assessment

medium, this changes the organization of a couple of widely shared components, it would be good to check that they're still sound.

### CODE REVIEW

File: app/src/organisms/ProtocolSetup/RunSetupCard/ModuleSetup/__tests__/ModuleSetup.test.tsx [hunk 1]
```diff
@@ -246,66 +244,31 @@ describe('ModuleSetup', () => {
...
-      moduleRenderCoords,
-    }
-
     const { getByText, getByRole } = render(props)
-    getByText('mock module viz magneticModuleType')
-    getByText('mock module viz thermocyclerModuleType')
     getByText('mock module info magneticModuleV2')
     const button = getByRole('button', { name: 'Proceed to Labware Setup' })
-    expect(button).not.toHaveAttribute('disabled')
+    expect(button).not.toBeDisabled()
```
Comment (shlokamin) [inline app/src/organisms/ProtocolSetup/RunSetupCard/ModuleSetup/__tests__/ModuleSetup.test.tsx:272]: didn't know `toBeDisabled()` was a thing, cool!

File: app/src/organisms/ProtocolSetup/RunSetupCard/LabwareSetup/index.tsx [hunk 2]
```diff
@@ -109,70 +97,63 @@ export const LabwareSetup = (props: LabwareSetupProps): JSX.Element | null => {
...
-            backgroundColor={C_BLUE}
-            id={'LabwareSetup_proceedToRunButton'}
-            {...linkProps}
-            {...targetProps}
-          >
-            {t('proceed_to_run')}
-          </PrimaryBtn>
-          {proceedToRunDisabled && (
-            <Tooltip {...tooltipProps}>{proceedToRunDisabledReason}</Tooltip>
-          )}
```
Comment (shlokamin) [inline app/src/organisms/ProtocolSetup/RunSetupCard/LabwareSetup/index.tsx:175]: nice, i like this better! didn't realize we had updated the designs to move the proceed to run button outside of the three accordions

File: app/src/organisms/ProtocolSetup/RunSetupCard/ModuleSetup/index.tsx [hunk 3]
```diff
@@ -44,51 +41,45 @@ const DECK_LAYER_BLOCKLIST = [
...
+  const attachedModules = useSelector((state: State) =>
+    getAttachedModules(state, robotName)
+  )
+
+  const moduleModels = map(
+    moduleRenderInfoById,
+    ({ moduleDef }) => moduleDef.model
+  )
+
+  const hasADuplicateModule = new Set(moduleModels).size !== moduleModels.length
```
Comment (shlokamin) [inline app/src/organisms/ProtocolSetup/RunSetupCard/ModuleSetup/index.tsx:76]: 😎

File: app/src/organisms/ProtocolSetup/RunSetupCard/__tests__/ProceedToRunCta.test.tsx [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+import * as hooks from '../hooks'
+import { ProceedToRunCta } from '../ProceedToRunCta'
+import { mockProtocolPipetteTipRackCalInfo } from '../../../../redux/pipettes/__fixtures__'
+import { getProtocolPipetteTipRackCalInfo } from '../../../../redux/pipettes'
+import type { ProtocolPipetteTipRackCalDataByMount } from '../../../../redux/pipettes/types'
+
+jest.mock('../hooks')
+jest.mock('../../../../redux/pipettes/__fixtures__')
+jest.mock('../../../../redux/pipettes')
+jest.mock('../../../../redux/pipettes/types')
```
Comment (shlokamin) [inline app/src/organisms/ProtocolSetup/RunSetupCard/__tests__/ProceedToRunCta.test.tsx]: how come you have to mock the types?

File: app/src/organisms/ProtocolSetup/RunSetupCard/__tests__/ProceedToRunCta.test.tsx [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    const button = getByRole('button', { name: 'Proceed to Run' })
+    expect(button).not.toHaveAttribute('disabled')
+  })
+  it('should be disabled with modules not connected tooltip when there are missing moduleIdd', () => {
+    mockUseMissingModuleIds.mockReturnValue(['temperatureModuleV1'])
+    const { getByRole } = render(props)
+    expect(typeof ProceedToRunCta).toBe('function')
+    const button = getByRole('button', { name: 'Proceed to Run' })
+    expect(button).toHaveAttribute('disabled')
+  })
```
Comment (shlokamin) [inline app/src/organisms/ProtocolSetup/RunSetupCard/__tests__/ProceedToRunCta.test.tsx]: these test names make it seem like we're asserting on the tooltip content,  but we're not — should we do that? Also, `moduleIdd` 😛

File: app/src/organisms/ProtocolSetup/RunSetupCard/__tests__/ProceedToRunCta.test.tsx [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+  it('should be disabled with modules not connected and calibration not completed tooltip if missing cal and moduleIds', () => {
+    mockUseMissingModuleIds.mockReturnValue(['temperatureModuleV1'])
+    mockGetProtocolPipetteTiprackData.mockReturnValue({
+      left: MOCK_PIPETTE_NOT_CALIBRATED,
+      right: null,
+    } as any)
+    const { getByRole } = render(props)
+    expect(typeof ProceedToRunCta).toBe('function')
```
Comment (shlokamin) [inline app/src/organisms/ProtocolSetup/RunSetupCard/__tests__/ProceedToRunCta.test.tsx]: how come we're asserting on the type of `ProceedToRunCta`?

File: app/src/organisms/ProtocolSetup/RunSetupCard/__tests__/hooks.test.tsx [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  type: 'magneticModuleType' as ModuleType,
+}
+
+describe('useMissingModuleIds', () => {
+  const store: Store<State> = createStore(jest.fn(), {})
+
+  beforeEach(() => {
+    store.dispatch = jest.fn()
+
+    mockGetConnectedRobot.mockReturnValue(mockConnectedRobot)
```
Comment (shlokamin) [inline app/src/organisms/ProtocolSetup/RunSetupCard/__tests__/hooks.test.tsx]: is there a reason we're sometimes using `when` mocks and sometimes not in this test?

File: app/src/organisms/ProtocolSetup/RunSetupCard/hooks.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    getAttachedModules(state, robot === null ? null : robot.name)
+  )
+  if (robot === null) {
+    return []
+  }
+
+  interface ModuleMatchResults {
+    missingModuleIds: string[]
+    remainingAttachedModules: AttachedModule[]
+  }
```
Comment (shlokamin) [inline app/src/organisms/ProtocolSetup/RunSetupCard/hooks.ts]: while it's totally valid to declare type definitions inside of a hook, it does seem a little odd... my style preference would be to move it out of the hook just to logically separate concerns

File: app/src/organisms/ProtocolSetup/RunSetupCard/hooks.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+            ...acc,
+            remainingAttachedModules: acc.remainingAttachedModules.filter(
+              (_m, i) => i !== moduleTypeMatchIndex
+            ),
+          }
+        : {
+            ...acc,
+            missingModuleIds: [...acc.missingModuleIds, id],
+          }
+    },
```
Comment (shlokamin) [inline app/src/organisms/ProtocolSetup/RunSetupCard/hooks.ts]: can some of the logic here be extracted into functions with semantic names to indicate what's going on here? it's a little hard to follow as is. also, can we use more explicit variable names than `_m` and `i`?

if all we're trying to do is determine if module ids exist in `moduleRenderInfoById` but not `attachedModules`, we can probably just use lodash's [difference array util](https://lodash.com/docs/4.17.15#difference)

**Review by shlokamin — APPROVED** [review]
left some comments we can address later, but great work @b-cooper! these module renders look really freakin' good 😍 

- [x] take a look through the code changes and call out anything suspicious
- [x] inspect the new protocol upload flow.  If the protocol loaded requires modules that are not attached to the current robot, the main "Proceed to Run" CTA should be disabled with an explanatory tooltip
- [x] inspect the new Module and Labware Setup steps, hifi Module svgs should render in both, though nested labware should only appear in the latter  
- [x] smoke test PD to ensure that it renders the deck with modules and labware just as it did before
- [x] smoke test protocol-library kludge to ensure that it's usage of legacy hardware simulation components is uneffected
- [x] take a look at the new stories associated with the `Module` component in the components library

File: app/src/organisms/ProtocolSetup/RunSetupCard/__tests__/ProceedToRunCta.test.tsx (hunk 4 above)
Comment (jerader) [reply app/src/organisms/ProtocolSetup/RunSetupCard/__tests__/ProceedToRunCta.test.tsx]: whoops, i'm the one who wrote that in my branch, which was merged into this one. i think that mock can be deleted

**Review by jerader — APPROVED** [review]
Very excited to see the new module visualizations! Looking great 🚀

File: app/src/organisms/ProtocolSetup/RunSetupCard/hooks.ts (hunk 9 above)
Comment (b-cooper) [reply app/src/organisms/ProtocolSetup/RunSetupCard/hooks.ts]: I'll clean it up a bit, it could definitely be more readable. 

Unless I'm missing something, I don't think the difference array util will work here. The attached modules from the robot come to us with their serial numbers as well as type/model information.  Whereas the modules in the protocol have a protocol specific unique instance identifier. What we're trying to do in this function is iterate over each of the required protocol modules and pick a match from the attached modules based on their model string alone. We're also ensuring that every attached module only gets matched to a single requested module.

Comment (shlokamin) [reply app/src/organisms/ProtocolSetup/RunSetupCard/hooks.ts]: ah, i see. yeah in that case just using the difference array util won't work. thx for explaining!
